ecr/manifest_writer_test.go

```diff
@@ -24,28 +24,136 @@ import (
+		MediaType: ocispec.MediaTypeImageManifest,
+	}
+	// root image Object has its digest appended.
+	imageObject := imageTag + "@" + imageDigest.String()
+	imageECRSpec := ECRSpec{
+		arn: arn.ARN{
+			AccountID: registry,
+		},
+		Repository: repository,
+		Object:     imageObject,
+	}
+	// root image ref is the ECRSpec's formatted ref, with the digest of the
+	// root descriptor. For a single manifest image, that's the manifest's
+	// digest.
+	refKey := imageECRSpec.Canonical()
+
+	t.Log("image Object: ", imageObject)
+	t.Log("image digest: ", imageDigest)
+
 	callCount := 0
 	client := &fakeECRClient{
 		PutImageFn: func(_ aws.Context, input *ecr.PutImageInput, _ ...request.Option) (*ecr.PutImageOutput, error) {
 			callCount++
+
 			assert.Equal(t, registry, aws.StringValue(input.RegistryId))
 			assert.Equal(t, repository, aws.StringValue(input.RepositoryName))
-			assert.Equal(t, imageTag, aws.StringValue(input.ImageTag))
+			assert.Equal(t, imageTag, aws.StringValue(input.ImageTag), "should use ref's tag")
 			assert.Equal(t, manifestContent, aws.StringValue(input.ImageManifest))
+
 			return &ecr.PutImageOutput{
-				Image: &ecr.Image{ImageId: &ecr.ImageIdentifier{ImageDigest: aws.String(imageDigest)}},
+				Image: &ecr.Image{
+					ImageId: &ecr.ImageIdentifier{
+						ImageTag:    input.ImageTag,
+						ImageDigest: aws.String(imageDigest.String()),
+					},
+				},
+			}, nil
+		},
+	}
+	mw := &manifestWriter{
+		desc: imageDesc,
+		base: &ecrBase{
+			client:  client,
+			ecrSpec: imageECRSpec,
+		},
+		tracker: docker.NewInMemoryTracker(),
+		ref:     refKey,
+		ctx:     context.Background(),
+	}
+
+	count, err := mw.Write([]byte(manifestContent[:3]))
+	require.NoError(t, err, "failed to write to manifest writer")
+	assert.Equal(t, 3, count, "wrong number of bytes")
+
+	count, err = mw.Write([]byte(manifestContent[3:]))
+	require.NoError(t, err, "failed to write to manifest writer")
+	assert.Equal(t, len(manifestContent)-3, count, "wrong number of bytes")
+
+	assert.Equal(t, 0, callCount, "PutImage should not be called until committed")
+
+	err = mw.Commit(context.Background(), int64(len(manifestContent)), imageDigest)
+	require.NoError(t, err, "failed to commit")
+	assert.Equal(t, 1, callCount, "PutImage should be called once")
+}
+
+func TestManifestWriterNoTagCommit(t *testing.T) {
+	const (
+		registry   = "registry"
+		repository = "repository"
+		imageTag   = "tag"
+
+		memberManifestContent = "manifest content"
+	)
+
+	// The root image, this is the target digest which is treated as an Image
+	// Index in this test case.
+	imageDigest := testdata.ImageDigest
+	// Image pushes include the root image digest in the object:
+	//
+	// ie: "latest@sha256:ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"
+	imageObject := imageTag + "@" + imageDigest.String()
+
+	// A member manifest that was listed and being "pushed" in the test case.
+	memberDesc := ocispec.Descriptor{
+		Digest:    testdata.GenerateDigest(),
```

Comment:
Do we need a generated digest here or can we use one of the other digests in `testdata`?

Reply:
I think this can be a placeholder. Will check and update if so.

Reply:
Ah, I see what's going on - this one is generated is to ensure the value is [distinct from the manifest list's descriptor](https://github.com/awslabs/amazon-ecr-containerd-resolver/pull/13/files/4c5a1a3529260906e0998bf1f5bcbf967ba8991c..f4c106f482fd7d8526488f1818cbafa3aec9971b#diff-08e31da21d34ea11b58e6c05528847fbR122) just above it. A different value is generated to verify that the digest written is for the image listed in `manifests` rather than the parent's (by way of guaranteeing a differing digest and comparing against it).

Reply:
Can you use a constant value like "member-manifest-digest"?

Reply:
Sure can! And will :)